### PR TITLE
Make flex_attention honor autocast (fixes silent fp32 attention, ~4x training slowdown)

### PR DESCRIPTION
**Pull request by shuaills**
## Summary

With the default training setup (mixed_precision: "bf16" + attn_implementation: "flex_attention"), the flex attention path can silently receive fp32 Q/K/V at the attention-kernel boundary and dispatch fp32 forward/backward kernels. On our H100 training setup this costs roughly 4× end-to-end throughput.

This PR restores bf16 flex attention in mixed precision with a 32-line repo-local change using Transformers' public AttentionInterface API.

## Mechanism
Accelerate bf16 mixed precision keeps model parameters in fp32.
In Qwen3 attention, per-head q_norm / k_norm can promote projected Q/K activations back to fp32.
scaled_dot_product_attention is covered by autocast, so the SDPA fallback path casts eligible attention inputs back to the active autocast dtype.
flex_attention is not covered in the same way in this stack, so fp32 inputs can pass through unchanged and dispatch fp32 flex kernels. As a result, the two attn_implementation paths can differ not only in backend but also in effective attention dtype.
At head_dim=128, the fp32 flex backward template is a register-pressure pathology: 61.8 ms vs 5.1 ms per layer-call on H100 with identical mask, shape, GQA layout, and gradient strides — about 12× at the kernel level and about 4× end-to-end.

## Fix

Register omnivoice_flex_attention via AttentionInterface.register. The registered implementation casts Q/K/V to the active autocast dtype at the flex-attention boundary when autocast is enabled and the inputs are still fp32.

This only controls the input dtype passed to the flex kernel; it does not force low-precision softmax accumulation. The behavior matches the intended mixed-precision boundary behavior of SDPA-style attention paths.

No global attention registry entries are modified. The change is a strict no-op when autocast is off or inputs are already low precision.
## Measured

Production fork of this repo, 0.6B Qwen3 backbone, packed 20480 tokens/GPU, H100:

| Metric | fp32 attention (current behavior) | bf16 cast (this PR) | Δ |
|---|---|---|---|
| flex backward kernel, per layer-call¹ | 61.8 ms | 5.1 ms | **12.1×** |
| training step time (3×H100) | 3.240 s/it | 0.755 s/it | **4.3×** |
| training throughput, tokens/s/GPU | 6,321 | 27,126 | 4.3× |
| MFU² | 3.3% | 14.3% | +11.0 pp |
| loss parity vs fp32 attention (300 steps, same seed) | — | worst-rel 0.43% | — |

¹ identical mask, shapes, GQA layout, and gradient strides; head_dim=128.
² model FLOPs from exact parameter census + real BlockMask visible-pair counts, against 989 TFLOPS dense bf16 peak.

We can attach a step-time timeline from a continuous production run crossing the fix boundary (same job/data/hardware before vs after) in a follow-up comment.

## Minimal repro

CPU-only, stock transformers ≥ 5.3 — shows fp32 arriving at the attention interface under bf16 autocast:

```python
import torch
from transformers import Qwen3Config, Qwen3Model
from transformers.modeling_utils import AttentionInterface

seen = {}
def probe(module, q, k, v, *a, **kw):
    seen["dtype"] = q.dtype
    return q.transpose(1, 2).contiguous(), None

AttentionInterface.register("probe", probe)
m = Qwen3Model(Qwen3Config(hidden_size=64, num_hidden_layers=1, num_attention_heads=4,
                           num_key_value_heads=2, head_dim=16, vocab_size=128))  # fp32 weights = master weights
m.set_attn_implementation("probe")
with torch.no_grad(), torch.autocast("cpu", dtype=torch.bfloat16):
    m(torch.randint(0, 128, (1, 8)))
print(seen["dtype"])  # torch.float32 — should be bfloat16
```

## Note

The root fix arguably belongs in PyTorch (`flex_attention` isn't in autocast's op coverage while SDPA is); this PR is the minimal repo-local correction until then.

### CODE REVIEW

**Comment by zhu-han** [comment]
Thanks a lot for this PR! Merged.
